backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt

```diff
@@ -0,0 +1,274 @@
+
+        logger.error("There is no <users> tag in external XML log")
+        return emptyList()
+    }
+
+    private fun parseContestTime(doc: Document): Pair<Duration, Duration> {
+        val config = doc.child(0)
+        val duration = config.attr("duration")
+        val fogTime = config.attr("fog_time")
+        return Pair(duration.toLong().seconds, (duration.toLong() - fogTime.toLong()).seconds)
+    }
+
+    private fun parseEjudgeTime(time: String): Instant {
+        val formattedTime = time
+            .replace("/", "-")
+            .replace(" ", "T")
+        return guessDatetimeFormat(formattedTime)
+    }
+
+    private fun parseContestInfo(element: Element, onRunChanges: (RunInfo) -> Unit) {
+        val dur = element.attr("duration").toLong()
+        val startTime = parseEjudgeTime(element.attr("start_time"))
+        val endTime = startTime.plus(dur.seconds)
+        val currentTime = parseEjudgeTime(element.attr("current_time"))
+
+        val status: ContestStatus = if (currentTime >= endTime) {
+            ContestStatus.OVER
+        } else if (currentTime < startTime) {
+            ContestStatus.BEFORE
+        } else {
+            ContestStatus.RUNNING
+        }
+
+        if (status == ContestStatus.RUNNING && contestData.status !== ContestStatus.RUNNING) {
+            contestData.startTime = startTime
+        }
+        contestData.status = status
+        contestData.contestTime = currentTime - startTime
+
+        element.children().forEach {
+            if ("runs" == it.tagName()) {
+                parseRuns(contestData, it, onRunChanges)
+            }
+        }
+    }
+
+    private fun parseRuns(
+        contestInfo: EjudgeContestInfo,
+        element: Element,
+        onRunChanges: (RunInfo) -> Unit
+    ) {
+        if (contestInfo.status == ContestStatus.BEFORE) {
+            return
+        }
+        element.children().forEach { run ->
+            parseRunInfo(contestInfo, run, onRunChanges)
+        }
+    }
+
+    private fun parseRunInfo(
+        contestInfo: EjudgeContestInfo,
+        element: Element,
+        onRunChanges: (RunInfo) -> Unit
+    ) {
+        val time = element.attr("time").toLong().seconds
+        if (time > contestInfo.contestTime) {
+            return
+        }
+
+        val teamSystemId = element.attr("user_id").toInt()
+        val teamId = contestInfo.teams[teamSystemId.toString()]!!.teamInfo.id
+        val runId = element.attr("run_id").toInt()
+
+        // Ejudge has 1-indexed problem numeration
+        val problemId = element.attr("prob_id").toInt() - 1
+        val isFrozen = time >= contestInfo.freezeTime
+        val oldRun = contestInfo.teams[teamSystemId.toString()]!!.runs[problemId].getOrDefault(runId, null)
+        val result = when {
+            isFrozen -> ""
+            else -> statusMap.getOrDefault(element.attr("status"), "WA")
+        }
+        val percentage = when {
+            isFrozen -> 0.0
+            "" == result -> 0.0
```

Comment:
I see a test field in xml? Does it exist when the run is in progress?

Reply:
Either it exists, or it equals to 0 when the run is in progress, I don't remember. But both of this cases are useless for this functionality.